zebra-script/src/lib.rs

```diff
@@ -198,31 +128,31 @@ impl CachedFfiTransaction {
             transparent::Input::Coinbase { .. } => Err(Error::TxCoinbase)?,
         };
 
-        let ctx = Box::new(SigHashContext {
-            input_index: n_in,
-            sighasher: SigHasher::new(&self.transaction, branch_id, &self.all_previous_outputs),
-        });
-        // SAFETY: The `script_*` fields are created from a valid Rust `slice`.
-        let ret = unsafe {
-            zcash_script::zcash_script_verify_callback(
-                (&*ctx as *const SigHashContext) as *const c_void,
-                Some(sighash),
-                lock_time,
-                is_final,
-                script_pub_key.as_ptr(),
-                script_pub_key.len() as u32,
-                signature_script.as_ptr(),
-                signature_script.len() as u32,
-                flags,
-                &mut err,
+        let calculate_sighash = |script_code: &[u8], hash_type: zcash_script::HashType| {
+            let script_code_vec = script_code.to_vec();
+            let mut our_hash_type = match hash_type.signed_outputs {
+                zcash_script::SignedOutputs::All => HashType::ALL,
+                zcash_script::SignedOutputs::Single => HashType::SINGLE,
+                zcash_script::SignedOutputs::None => HashType::NONE,
+            };
+            if hash_type.anyone_can_pay {
+                our_hash_type |= HashType::ANYONECANPAY;
+            }
+            Some(
+                SigHasher::new(&self.transaction, branch_id, &self.all_previous_outputs)
+                    .sighash(our_hash_type, Some((input_index, script_code_vec)))
+                    .0,
             )
         };
-
-        if ret == 1 {
-            Ok(())
-        } else {
-            Err(Error::from(err))
-        }
+        zcash_script::CxxRustComparisonInterpreter::verify_callback(
```

Comment:
Does the comparison interpreter always perform the comparisons, or is this controlled by feature flags?

Reply:
The comparison interpreter always performs the comparison. I think we would want a feature flag in Zebra to select the interpreter that’s used. Seems like a reasonable set of cases would be
- c++
- rust
- comparison
- stepwise comparison

Unfortunately, I can’t think of a comprehensible way to break it down into two orthogonal features. One clear dimension is whether or not it’s a comparison.

Reply:
Actually, a slightly different suggestion:

I think zcash_script should have a `rust_interpreter` feature (@ebfull had this in his draft, but I had removed it). This fits with the future organization, IMO, when the rust impl will live in zcash_transparent, and so the dependency on zcash_transparent would be conditional on the feature.

If that feature is disabled, then the C++ interpreter is the _only_ option.

Then in zebra … should it be a feature or configuration to select the particular interpreter (if the zcash_script `rust_interpreter` feature is disabled, then you’ll get the C++ interpreter no matter what the configuration says).